server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -1032,12 +1076,147 @@ public int hashCode() {
+            return Objects.hash(m, efConstruction, confidenceInterval);
+        }
+
+        @Override
+        public String toString() {
+            return "{type="
+                + type
+                + ", m="
+                + m
+                + ", ef_construction="
+                + efConstruction
+                + ", confidence_interval="
+                + confidenceInterval
+                + "}";
+        }
+
+        @Override
+        boolean supportsElementType(ElementType elementType) {
+            return elementType != ElementType.BYTE;
+        }
+
+        @Override
+        boolean updatableTo(IndexOptions update) {
+            return Objects.equals(this, update);
+        }
+
+        @Override
+        void validateDimension(int dim) {
+            if (dim % 2 != 0) {
+                throw new IllegalArgumentException("int4_hnsw only supports even dimensions; provided=" + dim);
+            }
+        }
+    }
+
+    static class Int4FlatIndexOptions extends IndexOptions {
+        private final float confidenceInterval;
+
+        Int4FlatIndexOptions(Float confidenceInterval) {
+            super("int4_flat");
+            // The default confidence interval for int4 is dynamic quantiles, this provides the best relevancy and is
+            // effectively required for int4 to behave well across a wide range of data.
+            this.confidenceInterval = confidenceInterval == null ? 0f : confidenceInterval;
+        }
+
+        @Override
+        public KnnVectorsFormat getVectorsFormat() {
+            return new ES813Int8FlatVectorFormat(confidenceInterval, 4, true);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field("type", type);
+            builder.field("confidence_interval", confidenceInterval);
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Int4FlatIndexOptions that = (Int4FlatIndexOptions) o;
+            return Objects.equals(confidenceInterval, that.confidenceInterval);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(confidenceInterval);
+        }
+
+        @Override
+        public String toString() {
+            return "{type=" + type + ", confidence_interval=" + confidenceInterval + "}";
+        }
+
+        @Override
+        boolean supportsElementType(ElementType elementType) {
+            return elementType != ElementType.BYTE;
+        }
+
+        @Override
+        boolean updatableTo(IndexOptions update) {
+            // TODO: add support for updating from flat, hnsw, and int8_hnsw and updating params
+            return Objects.equals(this, update);
```

Comment:
this would be nice to add. we might also do it in a separate PR.

Reply:
@tteofili I was thinking a separate PR due to the additional testing & code churn.